libsql/src/sync.rs

```diff
@@ -93,4 +121,175 @@ impl SyncContext {
     pub(crate) fn durable_frame_num(&self) -> u32 {
         self.durable_frame_num
     }
+
+    pub(crate) fn generation(&self) -> u32 {
+        self.generation
+    }
+
+    async fn write_metadata(&mut self) -> Result<()> {
+        let path = format!("{}-info", self.db_path);
+
+        let mut metadata = MetadataJson {
+            hash: 0,
+            version: METADATA_VERSION,
+            durable_frame_num: self.durable_frame_num,
+            generation: self.generation,
+        };
+
+        metadata.set_hash();
+
+        let contents = serde_json::to_vec(&metadata).unwrap();
+
+        atomic_write(path, &contents[..]).await.unwrap();
+
+        Ok(())
+    }
+
+    async fn read_metadata(&mut self) -> Result<()> {
+        let path = format!("{}-info", self.db_path);
+
+        if !std::fs::exists(&path).unwrap() {
+            tracing::debug!("no metadata info file found");
+            return Ok(());
+        }
+
+        let contents = tokio::fs::read(&path).await.unwrap();
+
+        let metadata = serde_json::from_slice::<MetadataJson>(&contents[..]).unwrap();
+
+        metadata.verify_hash()?;
+
+        // TODO(lucio): convert this into a proper error
+        assert_eq!(
```

Comment:
debug assert maybe?

Reply:
I think all in all we want this to panic in every release mode but I expect this to go away very shortly in a follow up.